src/workerd/api/container.c++

```diff
@@ -48,18 +48,64 @@ void Container::start(jsg::Lock& js, jsg::Optional<StartupOptions> maybeOptions)
   }
 
   IoContext::current().addTask(req.sendIgnoringResult());
-
   running = true;
+  monitorOnBackgroundIfNeeded();
+}
+
+void Container::monitorOnBackgroundIfNeeded() {
+  if (monitoring || !running) {
+    return;
+  }
+  monitoring = true;
+
+  if (monitorKjPromise == kj::none) {
+    monitorKjPromise = rpcClient->monitorRequest(capnp::MessageSize{4, 0})
+                           .send()
+                           .then([](auto&& results) -> uint8_t {
+      return results.getExitCode();
+    }).fork();
+  }
+
+  auto req = KJ_ASSERT_NONNULL(monitorKjPromise)
+                 .addBranch()
+                 .then([](uint8_t exitCode) {}, [this, self = JSG_THIS](kj::Exception&& error) {
+    // Check if there is an existing monitor request made by the user.
+    // If not, we can abort the current context and error.
+    if (!monitoringExplicitly) {
+      IoContext::current().abort(kj::mv(error));
+    }
+  }).attach(kj::defer([this, self = JSG_THIS]() {
+    running = false;
+    monitoring = false;
+  }));
+
+  IoContext::current().addTask(kj::mv(req));
 }
 
-jsg::Promise<void> Container::monitor(jsg::Lock& js) {
+jsg::MemoizedIdentity<jsg::Promise<void>>& Container::monitor(jsg::Lock& js) {
   JSG_REQUIRE(running, Error, "monitor() cannot be called on a container that is not running.");
 
-  return IoContext::current()
-      .awaitIo(js, rpcClient->monitorRequest(capnp::MessageSize{4, 0}).send())
-      .then(js, [this](jsg::Lock& js, capnp::Response<rpc::Container::MonitorResults> results) {
+  monitoringExplicitly = true;
+
+  KJ_IF_SOME(memoized, monitorJsPromise) {
+    return memoized;
+  }
+
+  if (monitorKjPromise == kj::none) {
+    auto responsePromise =
+        rpcClient->monitorRequest(capnp::MessageSize{4, 0})
+            .send()
+            .then([](capnp::Response<rpc::Container::MonitorResults>&& results) -> uint8_t {
+      return results.getExitCode();
+    });
+    monitorKjPromise = responsePromise.fork();
+  }
+
+  auto jsPromise = IoContext::current()
+                       .awaitIo(js, KJ_ASSERT_NONNULL(monitorKjPromise).addBranch())
```

Comment:
This is not a new problem in your PR, so I'm not asking you to fix it, but it occurs to me that this `awaitIo()` will block the DO from hibernating until the `monitor()` call completes. I guess this will be complicated to fix, as if we do want to allow hibernation, then we also need to make sure to wake the DO back up when the container shuts down. I think we'd actually have to lie to the DO, claiming that the container is still running, up until the point the DO calls `monitor()`, at which point we would immediately inform it of what actually happened. Ugh, kind of complicated. (Again, not for this PR.)